src/NextGen/Gateways/PayPalCommerce/payPalCommerceGateway.tsx

```diff
@@ -0,0 +1,295 @@
+                    hostedFieldType="number"
+                    options={{
+                        selector: "#card-number",
+                        placeholder: "4111 1111 1111 1111",
+                    }}
+                />
+
+                <Flex gap="10px">
+                    <PayPalHostedField
+                        id="expiration-date"
+                        className="givewp-fields"
+                        style={CUSTOM_FIELD_STYLE}
+                        hostedFieldType="expirationDate"
+                        options={{
+                            selector: "#expiration-date",
+                            placeholder: "MM/YYYY",
+                        }}
+                    />
+                    <PayPalHostedField
+                        id="cvv"
+                        className="card-field"
+                        style={CUSTOM_FIELD_STYLE}
+                        hostedFieldType="cvv"
+                        options={{
+                            selector: "#cvv",
+                            placeholder: "CVV",
+                            maskInput: true,
+                        }}
+                    />
+                </Flex>
+                <div style={{display: "flex", gap: '10px'}}>
+
+                </div>
+
+                <HoistHostedFieldContext />
+            </PayPalHostedFieldsProvider>
+        )
+    }
+
+    const payPalCommerceGateway: Gateway = {
+        id: 'paypal-commerce',
+        supportsRecurring: false,
+        supportsCurrency(currency: string): boolean {
+            return true;
+        },
+        initialize() {
+            payPalDonationsSettings = this.settings;
+        },
+        beforeCreatePayment: async function (values): Promise<object> {
+
+            if(payPalOrderId) { // If order ID already set by payment buttons then return early.
+                return {
+                    payPalOrderId: payPalOrderId
+                };
+            }
+
+            if(!validateHostedFields()) {
+                throw new Error('Invalid hosted fields');
+            }
+
+            return hostedField.cardFields
+                .submit({cardholderName: cardholderName,})
+                .then(async (data) => {
+                    await fetch( `${payPalDonationsSettings.ajaxUrl}?action=give_paypal_commerce_approve_order&order=` + data.orderId, {
+                        method: 'POST',
+                        body: getFormData(),
+                    } );
+                    return {...data, payPalOrderId: data.orderId}
+                })
+                .catch((err) => {
+                    console.log('paypal commerce error', err)
+                    throw new Error('paypal commerce error');
+                });
+        },
+        Fields() { // Can we get this.settings to be available here?
+            return (
+                <FormFieldsProvider>
+                    <PayPalScriptProvider options={payPalDonationsSettings.sdkOptions}>
```

Comment:
Would it be possible to add some kind of loading state that takes over paypal fields when we're redirected back from the paypal button and the form is hitting the server?

Reply:
Yeah, I think we can watch the form status for `isSubmitting` and modify the UI accordingly. I'll take a look.

Reply:
Resolved, see https://github.com/impress-org/givewp-next-gen/pull/139/commits/46559d76bbb188c7db05684c2aa06ce9216b2822.